I did some more work on this. You can check the "err" field for the error, since `getCachedLastError` is now (apparently) useless:

MongoClient connections throw MongoException:

```
scala> AlphaDAO.insert(List(alpha.copy(id = 0), alpha.copy(id = 1)))
res0: List[Option[Int]] = List(Some(0), Some(1))

scala> AlphaDAO.insert(List(alpha.copy(id = 1), alpha.copy(id = 2)))
com.novus.salat.dao.SalatInsertError:

    SalatDAO[Alpha,int](alpha_dao_spec): insert failed!

    Collection: alpha_dao_spec
    WriteConcern: WriteConcern { "getlasterror" : 1} / (Continue on error? false)
    Error: com.mongodb.MongoException$DuplicateKey: { "serverUsed" : "127.0.0.1:27017" , "ok" : 1 , "n" : 0 , "err" : "insertDocument :: caused by :: 11000 E11000 duplicate key error index: test_salat.alpha_dao_spec.$_id_  dup key: { : 1 }" , "code" : 11000}


FAILED TO INSERT ONE OR MORE OF YOUR DBOs (SEE Error FOR DETAILS):
{ "_typeHint" : "com.novus.salat.test.dao.Alpha" , "_id" : 1 , "beta" : [ { "_typeHint" : "com.novus.salat.test.dao.Delta" , "y" : "delta7a" , "z" : "delta7b"}]}
{ "_typeHint" : "com.novus.salat.test.dao.Alpha" , "_id" : 2 , "beta" : [ { "_typeHint" : "com.novus.salat.test.dao.Delta" , "y" : "delta7a" , "z" : "delta7b"}]}


  at com.novus.salat.dao.SalatDAO.insert(SalatDAO.scala:382)
  ... 43 elided
Caused by: com.mongodb.MongoException$DuplicateKey: { "serverUsed" : "127.0.0.1:27017" , "ok" : 1 , "n" : 0 , "err" : "insertDocument :: caused by :: 11000 E11000 duplicate key error index: test_salat.alpha_dao_spec.$_id_  dup key: { : 1 }" , "code" : 11000}
  at com.mongodb.CommandResult.getWriteException(CommandResult.java:88)
  at com.mongodb.CommandResult.getException(CommandResult.java:79)
  at com.mongodb.DBCollectionImpl.translateBulkWriteException(DBCollectionImpl.java:316)
  at com.mongodb.DBCollectionImpl.insert(DBCollectionImpl.java:191)
  at com.mongodb.DBCollectionImpl.insert(DBCollectionImpl.java:167)
  at com.mongodb.DBCollection.insert(DBCollection.java:93)
  at com.mongodb.casbah.MongoCollectionBase$class.insert(MongoCollection.scala:654)
  at com.mongodb.casbah.MongoCollection.insert(MongoCollection.scala:1162)
  at com.novus.salat.dao.SalatDAO.insert(SalatDAO.scala:383)
  ... 43 more
```

With legacy safety net: MongoConnection connections return a WriteResult with an "err" field that we can access via `getError`

```
scala> DeprecatedAlphaDAO.insert(List(alpha.copy(id = 1), alpha.copy(id = 2)))
com.novus.salat.dao.SalatInsertError:

    SalatDAO[Alpha,int](alpha_dao_spec): insert failed!

    Collection: alpha_dao_spec
    WriteConcern: WriteConcern { "getlasterror" : 1} / (Continue on error? false)
    Error: { "serverUsed" : "127.0.0.1:27017" , "connectionId" : 1098 , "err" : "insertDocument :: caused by :: 11000 E11000 duplicate key error index: test_salat.alpha_dao_spec.$_id_  dup key: { : 1 }" , "code" : 11000 , "n" : 0 , "ok" : 1.0}


FAILED TO INSERT ONE OR MORE OF YOUR DBOs (SEE Error FOR DETAILS):
{ "_typeHint" : "com.novus.salat.test.dao.Alpha" , "_id" : 1 , "beta" : [ { "_typeHint" : "com.novus.salat.test.dao.Delta" , "y" : "delta7a" , "z" : "delta7b"}]}
{ "_typeHint" : "com.novus.salat.test.dao.Alpha" , "_id" : 2 , "beta" : [ { "_typeHint" : "com.novus.salat.test.dao.Delta" , "y" : "delta7a" , "z" : "delta7b"}]}


  at com.novus.salat.dao.SalatInsertError$.apply(SalatDAOErrors.scala:80)
  at com.novus.salat.dao.SalatDAO$$anonfun$insert$3.apply(SalatDAO.scala:385)
  at com.novus.salat.dao.SalatDAO$$anonfun$insert$3.apply(SalatDAO.scala:385)
  at com.novus.salat.dao.SalatDAO.handleLegacyErrors(SalatDAO.scala:350)
  at com.novus.salat.dao.SalatDAO.insert(SalatDAO.scala:386)
  ... 43 elided
```

Because Mongo stops processing the list of DBOs at the first error (assuming you have halt on error enabled), it still might have inserted some records before hitting the error.  So, I made the exception text slightly more ambiguous on what DBOs failed when there are more than one (bulk insert).

--- a/salat-core/src/main/scala/com/novus/salat/dao/SalatDAO.scala
+++ b/salat-core/src/main/scala/com/novus/salat/dao/SalatDAO.scala
@@ -334,11 +334,20 @@ abstract class SalatDAO[ObjectType <: AnyRef, ID <: Any](val collection: MongoCo
    */
   override lazy val description = "SalatDAO[%s,%s](%s)".format(mot.runtimeClass.getSimpleName, mid.runtimeClass.getSimpleName, collection.name)
 
-  /** Legacy support. Code using MongoCollection instead of MongoClient might not throw MongoException */
-  private def handleLegacyErrors[T](wr: WriteResult)(failure: => Nothing)(success: => T) = {
+  /**
+   * Checks the "err" field in the write result, or the cached last error.
+   * "There should be no reason to use getError" - we found one.
+   */
+  private def defaultWriteResultErrorCheck(wr: WriteResult) = wr.getError != null || {
     val lastError = wr.getCachedLastError
-    if (lastError != null && !lastError.ok()) failure
-    else success
+    lastError != null && !lastError.ok()
+  }
+  /**
+   * Safety net for legacy code that is still using MongoConnection instead
+   * of MongoClient, and so will not reliably throw a MongoException.
+   */
+  private def handleLegacyErrors[T](wr: WriteResult, hasError: WriteResult => Boolean = defaultWriteResultErrorCheck _)(failure: => Nothing)(success: => T) = {
+    if (hasError(wr)) failure else success
   }
 
   /**

--- a/salat-core/src/main/scala/com/novus/salat/dao/SalatDAOErrors.scala
+++ b/salat-core/src/main/scala/com/novus/salat/dao/SalatDAOErrors.scala
@@ -36,9 +36,19 @@ protected[dao] object SalatDAOError {
   implicit class SomeKindOfMongoError(val cause: SalatDAOMongoError) extends AnyVal {
     def toErrorString: String = cause.fold({ wr => s"$wr" }, { ex => s"$ex" })
   }
+
+  private val MulitDboFailureMsg = "ONE OR MORE OF YOUR DBOs FAILED"
+
+  def dboFailures(dbos: List[DBObject], opThatFailed: String) = {
+    val single = dbos.size == 1
+    s"""
+       |FAILED TO ${opThatFailed.toUpperCase()} ${if (single) "DBO" else s"ONE OR MORE OF YOUR DBOs (SEE Error FOR DETAILS)"}:
+       |${if (single) dbos.head else dbos.mkString("\n")}
+     """.stripMargin
+  }
 }
 
-import SalatDAOError.{SalatDAOMongoError, SomeKindOfMongoError}
+import SalatDAOError._
 
 abstract class SalatDAOError(
   whichDAO:        String,
@@ -55,9 +65,7 @@ abstract class SalatDAOError(
     WriteConcern: $wc
     Error: ${cause.toErrorString}
 
-    FAILED TO ${thingThatFailed.toUpperCase} ${if (dbos.size == 1) "DBO" else "DBOs"}
-    ${if (dbos.size == 1) dbos.head else dbos.mkString("\n")}
-
+    ${dboFailures(dbos, thingThatFailed)}
  """, cause.right.toOption.orNull)
 
 object SalatInsertError {

--- a/salat-core/src/main/scala/com/novus/salat/dao/SalatDAOErrors.scala
+++ b/salat-core/src/main/scala/com/novus/salat/dao/SalatDAOErrors.scala
@@ -31,9 +31,9 @@ import com.mongodb.casbah.TypeImports._
 import com.mongodb.{DBObject, WriteConcern}
 
 protected[dao] object SalatDAOError {
-  type SalatDAOMongoError = Either[WriteResult, MongoException]
+  type LegacyErrorOrMongoException = Either[WriteResult, MongoException]
 
-  implicit class SomeKindOfMongoError(val cause: SalatDAOMongoError) extends AnyVal {
+  implicit class SomeKindOfMongoError(val cause: LegacyErrorOrMongoException) extends AnyVal {
     def toErrorString: String = cause.fold({ wr => s"$wr" }, { ex => s"$ex" })
   }
 
@@ -55,7 +55,7 @@ abstract class SalatDAOError(
   thingThatFailed: String,
   collection:      MongoCollection,
   wc:              WriteConcern,
-  cause:           SalatDAOMongoError,
+  cause:           LegacyErrorOrMongoException,
   dbos:            List[DBObject]
 ) extends RuntimeException(s"""
 
@@ -69,7 +69,7 @@ abstract class SalatDAOError(
  """, cause.right.toOption.orNull)
 
 object SalatInsertError {
-  @deprecated("Use MongoClient instead of MongoCollection", "1.10.0")
+  @deprecated("Use MongoClient instead of MongoConnection", "1.10.0")
   def apply(
     description: String,
     collection:  MongoCollection,
@@ -84,12 +84,12 @@ case class SalatInsertError(
   description: String,
   collection:  MongoCollection,
   wc:          WriteConcern,
-  cause:       SalatDAOMongoError,
+  cause:       LegacyErrorOrMongoException,
   dbos:        List[DBObject]
 ) extends SalatDAOError(description, "insert", collection, wc, cause, dbos)
 
 object SalatRemoveError {
-  @deprecated("Use MongoClient instead of MongoCollection", "1.10.0")
+  @deprecated("Use MongoClient instead of MongoConnection", "1.10.0")
   def apply(
     description: String,
     collection:  MongoCollection,
@@ -104,12 +104,12 @@ case class SalatRemoveError(
   description: String,
   collection:  MongoCollection,
   wc:          WriteConcern,
-  cause:       SalatDAOMongoError,
+  cause:       LegacyErrorOrMongoException,
   dbos:        List[DBObject]
 ) extends SalatDAOError(description, "remove", collection, wc, cause, dbos)
 
 object SalatSaveError {
-  @deprecated("Use MongoClient instead of MongoCollection", "1.10.0")
+  @deprecated("Use MongoClient instead of MongoConnection", "1.10.0")
   def apply(
     description: String,
     collection:  MongoCollection,
@@ -124,7 +124,7 @@ case class SalatSaveError(
   description: String,
   collection:  MongoCollection,
   wc:          WriteConcern,
-  cause:       SalatDAOMongoError,
+  cause:       LegacyErrorOrMongoException,
   dbos:        List[DBObject]
 ) extends SalatDAOError(description, "save", collection, wc, cause, dbos)
 
@@ -134,7 +134,7 @@ abstract class SalatDAOQueryError(
   collection:      MongoCollection,
   query:           DBObject,
   wc:              WriteConcern,
-  cause:           SalatDAOMongoError
+  cause:           LegacyErrorOrMongoException
 ) extends RuntimeException(s"""
 
     $whichDAO: $thingThatFailed failed!
@@ -164,7 +164,7 @@ case class SalatRemoveQueryError(
   collection: MongoCollection,
   query:      DBObject,
   wc:         WriteConcern,
-  cause:      SalatDAOMongoError
+  cause:      LegacyErrorOrMongoException
 ) extends SalatDAOQueryError(whichDAO, "remove", collection, query, wc, cause)
 
 object SalatDAOUpdateError {
@@ -188,7 +188,7 @@ case class SalatDAOUpdateError(
   query:      DBObject,
   o:          DBObject,
   wc:         WriteConcern,
-  cause:      SalatDAOMongoError,
+  cause:      LegacyErrorOrMongoException,
   upsert:     Boolean,
   multi:      Boolean
 ) extends RuntimeException(s"""

--- a/salat-core/src/test/scala/com/novus/salat/test/dao/SalatDAOSpec.scala
+++ b/salat-core/src/test/scala/com/novus/salat/test/dao/SalatDAOSpec.scala
@@ -94,14 +94,14 @@ class SalatDAOSpec extends SalatSpec {
       AlphaDAO.insert() must_== Nil
     }
 
-    "handle MongoExceptions due to inserting records having duplicate keys" in new alphaContext {
+    "handle MongoExceptions when inserting an object having a duplicate key" in new alphaContext {
       AlphaDAO.insert(alpha7)
       AlphaDAO.insert(alpha7) must throwA[SalatInsertError].like {
         case ex: SalatInsertError => ex.getCause must beAnInstanceOf[DuplicateKeyException]
       }
     }
 
-    "handle MongoExceptions for bulk inserts having a duplicate key" in new alphaContext {
+    "handle MongoExceptions for bulk inserts where an object has a duplicate key" in new alphaContext {
       AlphaDAO.insert(alpha7)
 
       // Note: Mongo stops processing at the first dup in the list
@@ -114,12 +114,12 @@ class SalatDAOSpec extends SalatSpec {
       }
     }
 
-    "handle legacy errors due to inserting records having duplicate keys" in new alphaContext {
+    "handle legacy errors when inserting an object having a duplicate key" in new alphaContext {
       DeprecatedAlphaDAO.insert(alpha7)
       DeprecatedAlphaDAO.insert(alpha7) must throwA[SalatInsertError]
     }
 
-    "handle legacy errors for bulk inserts having a duplicate key" in new alphaContext {
+    "handle legacy errors for bulk inserts where an object has a duplicate key" in new alphaContext {
       DeprecatedAlphaDAO.insert(alpha7)
 
       // Note: Mongo stops processing at the first dup in the list
@@ -137,7 +137,7 @@ class SalatDAOSpec extends SalatSpec {
       }
     }
 
-    "handle legacy for invalid updates" in new alphaContext {
+    "handle legacy errors for invalid updates" in new alphaContext {
       DeprecatedAlphaDAO.insert(alpha7)
       DeprecatedAlphaDAO.update(MongoDBObject.empty, MongoDBObject("_id" -> 1)) must throwA[SalatDAOUpdateError]
     }

--- a/salat-core/src/test/scala/com/novus/salat/test/dao/SalatDAOSpec.scala
+++ b/salat-core/src/test/scala/com/novus/salat/test/dao/SalatDAOSpec.scala
@@ -29,9 +29,9 @@ package com.novus.salat.test.dao
 
 import com.mongodb.casbah.Imports._
 import com.mongodb.casbah.commons.MongoDBObject
-import com.mongodb.DuplicateKeyException
+import com.mongodb.{DuplicateKeyException, WriteConcernException}
 import com.novus.salat._
-import com.novus.salat.dao.SalatInsertError
+import com.novus.salat.dao.{SalatDAOUpdateError, SalatInsertError}
 import com.novus.salat.test._
 import com.novus.salat.test.global._
 import org.specs2.specification.Scope
@@ -94,17 +94,53 @@ class SalatDAOSpec extends SalatSpec {
       AlphaDAO.insert() must_== Nil
     }
 
-    "handle MongoExceptions for duplicate key inserts" in new alphaContext {
+    "handle MongoExceptions due to inserting records having duplicate keys" in new alphaContext {
       AlphaDAO.insert(alpha7)
       AlphaDAO.insert(alpha7) must throwA[SalatInsertError].like {
         case ex: SalatInsertError => ex.getCause must beAnInstanceOf[DuplicateKeyException]
       }
     }
 
-    "handle DAOs using deprecated MongoConnection, throwing an exception on a WriteResult error" in new alphaContext {
+    "handle MongoExceptions for bulk inserts having a duplicate key" in new alphaContext {
+      AlphaDAO.insert(alpha7)
+
+      // Note: Mongo stops processing at the first dup in the list
+      AlphaDAO.insert(List(alpha7, alpha1, alpha2, alpha3)) must throwA[SalatInsertError].like {
+        case ex: SalatInsertError =>
+          val mustBeMongoException = ex.getCause must beAnInstanceOf[DuplicateKeyException]
+          val onlyOneRecord = AlphaDAO.collection.count() must_== 1
+
+          mustBeMongoException and onlyOneRecord
+      }
+    }
+
+    "handle legacy errors due to inserting records having duplicate keys" in new alphaContext {
       DeprecatedAlphaDAO.insert(alpha7)
       DeprecatedAlphaDAO.insert(alpha7) must throwA[SalatInsertError]
-    }.pendingUntilFixed("duplicate insert fails silently when SalatDAO uses a MongoConnection instead of a MongoClient?!?")
+    }
+
+    "handle legacy errors for bulk inserts having a duplicate key" in new alphaContext {
+      DeprecatedAlphaDAO.insert(alpha7)
+
+      // Note: Mongo stops processing at the first dup in the list
+      val errorMustOccur = DeprecatedAlphaDAO.insert(List(alpha7, alpha1, alpha2, alpha3)) must throwA[SalatInsertError]
+      val onlyOneRecord = DeprecatedAlphaDAO.collection.count() must_== 1
+
+      errorMustOccur and onlyOneRecord
+    }
+
+    "handle MongoExceptions for invalid updates" in new alphaContext {
+      AlphaDAO.insert(alpha7)
+      AlphaDAO.update(MongoDBObject.empty, MongoDBObject("_id" -> 1)) must throwA[SalatDAOUpdateError].like {
+        case ex: SalatDAOUpdateError =>
+          ex.getCause must beAnInstanceOf[WriteConcernException]
+      }
+    }
+
+    "handle legacy for invalid updates" in new alphaContext {
+      DeprecatedAlphaDAO.insert(alpha7)
+      DeprecatedAlphaDAO.update(MongoDBObject.empty, MongoDBObject("_id" -> 1)) must throwA[SalatDAOUpdateError]
+    }
 
     "support findOne returning Option[T]" in new alphaContext {
       val _ids = AlphaDAO.insert(alpha4, alpha5, alpha6)

